codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/transformers/ConstrainedMemberTransform.kt

```diff
@@ -0,0 +1,210 @@
+            (0..100).forEach {
+                val extractedStructName = if (it == 0) makeStructName("") else makeStructName("$it")
+                if (structNameIsUnique(extractedStructName))
+                    return extractedStructName
+            }
+
+            throw IllegalStateException("A unique name for the overridden structure type could not be generated")
+        }
+
+        return generateUniqueName()
+    }
+
+    /**
+     * Returns the transformation that would be required to turn the given member shape
+     * into a non-constrained member shape.
+     */
+    private fun MemberShape.makeNonConstrained(
+        model: Model,
+        additionalNames: MutableSet<ShapeId>,
+    ): MemberShapeTransformation? {
+        val (constraintTraits, otherTraits) = this.allTraits.values
+            .partition {
+                memberConstraintTraitsToOverride.contains(it.javaClass)
+            }
+
+        // No transformation required in case the given MemberShape has no constraints.
+        if (constraintTraits.isEmpty())
+            return null
+
+        val targetShape = model.expectShape(this.target)
+        if (targetShape !is ToSmithyBuilder<*>)
+            UNREACHABLE("member target shapes will always be buildable")
+
+        when (val builder = targetShape.toBuilder()) {
+            is AbstractShapeBuilder<*, *> -> {
+                // Use the target builder to create a new standalone shape that would
+                // be added to the model later on. Keep all existing traits on the target
+                // but replace the ones that are overridden on the member shapes.
+                val existingNonOverriddenTraits =
+                    builder.allTraits.values.filter { existingTrait ->
+                        constraintTraits.none { it.toShapeId() == existingTrait.toShapeId() }
+                    }
+
+                val newTraits =
+                    existingNonOverriddenTraits + constraintTraits + SyntheticStructureFromConstrainedMemberTrait(
+                        this.id,
+                    )
+
+                // Create a new standalone shape that will be added to the model later on
+                val shapeId = overriddenShapeId(model, additionalNames, this.id)
+                val standaloneShape = builder.id(shapeId)
+                    .traits(newTraits)
+                    .build()
+
+                // Since the new shape has not been added to the model as yet, the current
+                // memberShape's target cannot be changed to the new shape.
+                return MemberShapeTransformation(standaloneShape, this, otherTraits)
+            }
+
+            else -> throw IllegalStateException("Constraint traits cannot to applied on ${this.id}") // FZ confirm how we are throwing exceptions
+        }
+
+        throw IllegalStateException("Constraint traits can only be applied to buildable types. ${this.id} is not buildable") // FZ confirm how we are throwing exceptions
```

Comment:
This line is not necessary. We must go through the first arm of the `when`, which `return`s.

You can rewrite this so that it's more apparent as follows, note the `return when`, each arm returns an expression:

```kotlin
return when (val builder = targetShape.toBuilder()) {
    is AbstractShapeBuilder<*, *> -> {
        ...
        MemberShapeTransformation(standaloneShape, this, otherTraits)
    }

    else -> UNREACHABLE("Constraint traits cannot to applied on ${this.id}")
}
```